<a href="https://colab.research.google.com/github/alloe16/NLP/blob/main/squad_bert.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#### Context Question Answering

In [1]:
!python -m deeppavlov riseapi squad_bert -p 5005


/usr/bin/python3: No module named deeppavlov


In [ ]:
!python -m deeppavlov riseapi squad_bert -p 1221

In [ ]:
!pip install -q deeppavlov
!python -m deeppavlov install squad_bert
from deeppavlov import build_model, configs

     |████████████████████████████████| 441 kB 40.0 MB/s 
     |████████████████████████████████| 54 kB 2.8 MB/s 
     |████████████████████████████████| 59 kB 6.9 MB/s 
     |████████████████████████████████| 57 kB 5.6 MB/s 
     |████████████████████████████████| 55 kB 3.7 MB/s 
     |████████████████████████████████| 42 kB 869 kB/s 
     |████████████████████████████████| 145 kB 78.5 MB/s 
     |████████████████████████████████| 8.2 MB 47.5 MB/s 
     |████████████████████████████████| 63 kB 543 kB/s 
     |████████████████████████████████| 80 kB 8.8 MB/s 
     |████████████████████████████████| 58 kB 5.0 MB/s 
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 4.4 MB 36.8 MB/s 
     |████████████████████████████████| 6.6 MB 65.0 MB/s 
     |████████████████████████████████| 182 kB 84.1 MB/s 
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
model_qa = build_model(configs.squad.squad_bert, download=True)

2022-12-02 06:41:11.382 INFO in 'deeppavlov.core.data.utils'['utils'] at line 95: Downloading from http://files.deeppavlov.ai/v1/squad/squad_torch_bert_cased.tar.gz to /root/.deeppavlov/models/squad_torch_bert/cased/squad_torch_bert_cased.tar.gz
INFO:deeppavlov.core.data.utils:Downloading from http://files.deeppavlov.ai/v1/squad/squad_torch_bert_cased.tar.gz to /root/.deeppavlov/models/squad_torch_bert/cased/squad_torch_bert_cased.tar.gz
100%|██████████| 1.18G/1.18G [01:35<00:00, 12.3MB/s]
2022-12-02 06:42:48.319 INFO in 'deeppavlov.core.data.utils'['utils'] at line 276: Extracting /root/.deeppavlov/models/squad_torch_bert/cased/squad_torch_bert_cased.tar.gz archive into /root/.deeppavlov/models/squad_torch_bert/cased/bert-base-cased
INFO:deeppavlov.core.data.utils:Extracting /root/.deeppavlov/models/squad_torch_bert/cased/squad_torch_bert_cased.tar.gz archive into /root/.deeppavlov/models/squad_torch_bert/cased/bert-base-cased


Downloading:   0%|          | 0.00/29.0 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/570 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/208k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/426k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/416M [00:00<?, ?B/s]

Some weights of the model checkpoint at bert-base-cased were not used when initializing BertForQuestionAnswering: ['cls.predictions.transform.dense.weight', 'cls.seq_relationship.bias', 'cls.predictions.transform.LayerNorm.bias', 'cls.predictions.decoder.weight', 'cls.predictions.transform.LayerNorm.weight', 'cls.seq_relationship.weight', 'cls.predictions.bias', 'cls.predictions.transform.dense.bias']
- This IS expected if you are initializing BertForQuestionAnswering from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForQuestionAnswering from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of BertForQuestionAnswering were not initialized from the model checkpoint at bert-base-cased and a

In [ ]:
model_qa(["In meteorology, precipitation is any product of the condensation of atmospheric water vapor \
         that falls under gravity. The main forms of precipitation include drizzle, rain, sleet, snow, \
         graupel and hail. Precipitation forms as smaller droplets coalesce via collision with other rain drops \
         or ice crystals within a cloud. Short, intense periods of rain in scattered locations are called showers."],
         ["Where do water droplets collide with ice crystals to form precipitation?"])

[['within a cloud'], [332], [1.0]]

In [ ]:
model_qa(["Prince Harry and fiancee American actress Meghan Markle have released more details about their May 19 wedding, revealing that the event will include a carriage ride through Windsor so they can share the big day with the public. \
The couple will marry at noon in St. George’s Chapel, the 15th century church on the grounds of Windsor Castle that has long been the backdrop of choice for royal occasions. Harry’s grandmother, Queen Elizabeth II, gave permission for use of the venue and will attend the wedding.\
 Kensington Palace said in a statement that the couple is “hugely grateful” for the many good wishes they have received and they hope the carriage ride will give the general public a chance to take part."],
 ["Who is going to marry?"])

[['Prince Harry and fiancee American actress Meghan Markle'],
 [0],
 [0.9999999403953552]]

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/deeppavlov/DeepPavlov/blob/master/docs/features/models/SQuAD.ipynb)

> Блок с отступами



[![Medium](https://img.shields.io/badge/Medium-12100E?style=for-the-badge&logo=medium&logoColor=white)](https://medium.com/deeppavlov/developing-qa-systems-for-any-language-with-deeppavlov-a9033d5231a8)

# Table of contents 

1. [Introduction to the task](#1.-Introduction-to-the-task)

2. [Get started with the model](#2.-Get-started-with-the-model)

3. [Use the model for prediction](#3.-Use-the-model-for-prediction)

    3.1. [Predict using Python](#3.1-Predict-using-Python)
    
    3.2. [Predict using CLI](#3.2-Predict-using-CLI)
     
4. [Train the model on your data](#4.-Train-the-model-on-your-data)
    
    4.1. [from Python](#4.1-Train-your-model-from-Python)
    
    4.2. [from CLI](#4.2-Train-your-model-from-CLI)
    
5. [Evaluate](#5.-Evaluate)
    
    5.1. [from Python](#5.1-Evaluate-from-Python)
    
    5.2. [from CLI](#5.2-Evaluate-from-CLI)
    
6. [Models list](#6.-Models-list)

# 1. Introduction to the task

Context Question Answering is a task of finding a fragment with an answer to a question in a given segment of context.

**Context**:

```
In meteorology, precipitation is any product of the condensation 
of atmospheric water vapor that falls under gravity. The main forms 
of precipitation include drizzle, rain, sleet, snow, graupel and hail… 
Precipitation forms as smaller droplets coalesce via collision with 
other rain drops or ice crystals within a cloud. Short, intense periods 
of rain in scattered locations are called “showers”.
```

**Question**:
```
Where do water droplets collide with ice crystals to form precipitation?
```

**Answer**: 
```
within a cloud
```

Datasets that follow this task format:

- [Stanford Question Answering Dataset (SQuAD) (EN)](https://rajpurkar.github.io/SQuAD-explorer/)

- [SberQuAD (RU)](https://paperswithcode.com/dataset/sberquad)

# 2. Get started with the model

First make sure you have the DeepPavlov Library installed.
[More info about the first installation](http://docs.deeppavlov.ai/en/master/intro/installation.html)

In [ ]:
!pip install --q deeppavlov

Then make sure that all the required packages for the model are installed.

In [ ]:
!python -m deeppavlov install squad_bert

`squad_torch_bert` here is the name of the model's *config_file*. [What is a Config File?](http://docs.deeppavlov.ai/en/master/intro/configuration.html) 

Configuration file defines the model and describes its hyperparameters. To use another model, change the name of the *config_file* here and further.
The full list of the models with their config names can be found in the [table](#6.-Models-list).


# 3. Use the model for prediction

## 3.1 Predict using Python

After [installing](#2.-Get-started-with-the-model) the model, build it from the config and predict.

In [ ]:
from deeppavlov import build_model

model = build_model('squad_bert', download=True)

**Input**: List[context, question]

**Output**: List[answer, start_character, logit]

In [ ]:
model(['DeepPavlov is a library for NLP and dialog systems.'], ['What is DeepPavlov?'])

[['a library for NLP and dialog systems'], [14], [200928.390625]]

## 3.2 Predict using CLI

You can also get predictions in an interactive mode through CLI.

In [ ]:
!python -m deeppavlov interact squad_bert -d

`-d` is an optional download key (alternative to `download=True` in Python code). The key `-d` is used to download the pre-trained model along with embeddings and all other files needed to run the model. 

Or make predictions for samples from *stdin*.

In [ ]:
!python -m deeppavlov predict squad_bert -f <file-name>

# 4. Train the model on your data


## 4.1 Train your model from Python

### Provide your data path

To train the model on your data, you need to change the path to the training data in the *config_file*. 

Parse the *config_file* and change the path to your data from Python.

In [ ]:
from deeppavlov import train_model
from deeppavlov.core.commands.utils import parse_config

model_config = parse_config('squad_bert')

#  dataset that the model was trained on
print(model_config['dataset_reader']['data_path'])

~/.deeppavlov/downloads/squad/


Provide a *data_path* to your own dataset. 

In [ ]:
# download and unzip a new example dataset
!wget http://files.deeppavlov.ai/datasets/squad-v1.1.tar.gz
!tar -xzvf "squad-v1.1.tar.gz"

Note that if you want to provide your own dataset, it should have the same format as the SQuAD dataset downloaded in this cell.

In [ ]:
# provide a path to the train file
model_config["dataset_reader"]["data_path"] = "/contents/train-v1.1.json"

### SQuAD dataset info

There are *two* versions of the SQuAD dataset available for training at the moment: 

- [SQuAD 1.1](https://arxiv.org/abs/1606.05250) contains 107,785 question-answer pairs on 536 articles. Dataset size: `33.52 MiB`.

- [SQuAD 2.0](https://arxiv.org/abs/1806.03822) combines all of the questions from SQuAD 1.1 with over 50,000 un-answerable questions written adversarially by crowdworkers. Dataset size: `44.34 MiB`.

### Train the model using new config

In [ ]:
model = train_model(model_config)

Use your model for prediction.

In [ ]:
model(['DeepPavlov is a library for NLP and dialog systems.'], ['What is DeepPavlov?'])

[['a library for NLP and dialog systems'], [14], [200928.390625]]

## 4.2 Train your model from CLI

In [ ]:
!python -m deeppavlov train squad_bert

# 5. Evaluate 

## 5.1 Evaluate from Python

In [ ]:
from deeppavlov import evaluate_model

model = evaluate_model('squad_bert', download=True)

## 5.1 Evaluate from CLI

In [ ]:
! python -m deeppavlov evaluate squad_bert -d

# 6. Models list

The table presents a list of all of the Context Question Answering models available in DeepPavlov Library.

| Config name  | Dataset | Language | Model Size | F1 score | EM  |
| :--- | --- | --- | --- | --- | ---: |
| squad_bert | SQuAD v1.1 | En | 1.3 GB | 88.86 | 81.49 |
| qa_squad2_bert | SQuAD v2.0 | En | 1.3 GB | 83.56 | 75.54 |
| qa_multisberquad_bert | MultiSQuAD | Multi | 2 GB | 80.76 | 63.81 |
| squad_ru_bert | SberQuAD | Ru | 2.0 GB | 84.71 | 66.21 |
| squad_ru_convers_distilrubert_2L | SberQuAD | Ru | 1.2 GB | 65.20 | 44.52 |
| squad_ru_convers_distilrubert_6L | SberQuAD | Ru | 1.6 GB | 80.57 | 61.54 | 
